### PR TITLE
Perform builtins injection for WORKSPACE-loaded bzl files.

### PR DESCRIPTION
**Pull request by benjaminp**
https://github.com/bazelbuild/bazel/issues/17713

### CODE REVIEW

**Comment by comius** [comment]
cc @Wyverald @meteorcloudy @brandjon

File: src/main/java/com/google/devtools/build/lib/packages/BazelStarlarkEnvironment.java
Comment (comius) [inline src/main/java/com/google/devtools/build/lib/packages/BazelStarlarkEnvironment.java:379]: Using `ruleClassProvider.getEnvironment()` looks like a semantic change. Before it was `uninjectedBuildBzlEnv`.

File: src/main/java/com/google/devtools/build/lib/skyframe/BzlLoadFunction.java
Comment (comius) [inline src/main/java/com/google/devtools/build/lib/skyframe/BzlLoadFunction.java:589]: There are 4 keys (also for builtins and for bzlmod). Does this condition work if it's changed to `key.isBuiltins()`? That would be slightly more simple.

File: src/main/java/com/google/devtools/build/lib/skyframe/RepositoryMappingFunction.java
Comment (comius) [inline src/main/java/com/google/devtools/build/lib/skyframe/RepositoryMappingFunction.java:139]: What is the justification for this check here?

The complexity of the code here is going slightly past my review capabilities and my first thought is this is just making the tests to pass and the code even more complex.

Similar complexity to this one is happening in BzlloadFunction.getRepositoryMapping, which I imagine is calling the location here.

Is there a way to simplify it?

File: src/main/java/com/google/devtools/build/lib/packages/BazelStarlarkEnvironment.java
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/packages/BazelStarlarkEnvironment.java:80]: `bzlmodBzlEnv` can actually just be the same as `workspaceBzlEnv`; that TODO was left before we had a full understanding of what .bzl environments were needed.

We can either merge the two in this PR, or clean up that TODO later separately.

File: src/main/java/com/google/devtools/build/lib/skyframe/RepositoryMappingFunction.java
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/skyframe/RepositoryMappingFunction.java:139]: I also don't understand this check. `@bazel_tools` should have its repo mapping calculated just like any other Bazel module, not the `VALUE_FOR_ROOT_MODULE_WITHOUT_REPOS` sentinel value.

> Similar complexity to this one is happening in BzlloadFunction.getRepositoryMapping, which I imagine is calling the location here.

Indeed, `BzlLoadFunction#getRepositoryMapping` calls to `RepositoryMappingFunction`, but the former is more complex because repo mapping computation for .bzl files has extra considerations due to some "partial" loading. `RepositoryMappingFunction` deals with the general case of "all WORKSPACE/Bzlmod/injection has finished and we're just loading a normal BUILD file (and similar environments) now".

Comment (benjaminp) [reply src/main/java/com/google/devtools/build/lib/skyframe/RepositoryMappingFunction.java:139]: This check avoids a fundamental cycle when bzlmod is enabled: workspace loading -> `@_builtins` loading -> `@_builtins` repository mapping -> `@bazel_tools` repository mapping -> `//external` package -> workspace loading.

The effect of this check, I believe, is you can't declare a repo_mapping for `@bazel_tools` in WORKSPACE. That's fine, I think? I'm certainly open to any other ideas on how to resolve this cycle.

File: src/main/java/com/google/devtools/build/lib/packages/BazelStarlarkEnvironment.java
Comment (benjaminp) [reply src/main/java/com/google/devtools/build/lib/packages/BazelStarlarkEnvironment.java:382]: No, that is about having the same `native` object for WORKSPACE-loaded and BUILD-loaded bzls. This change maintains the distinction.

File: src/main/java/com/google/devtools/build/lib/skyframe/RepositoryMappingFunction.java
Comment (comius) [reply src/main/java/com/google/devtools/build/lib/skyframe/RepositoryMappingFunction.java:139]: The conditions that lead to this line are (I took the liberty to enumerate them using and IDE, because I find it hard to follow {,} on github):
- enableBzlMod 
- not @_builtins
- not (main repo && rootModuleShouldSeeWorkspaceRepos)
- bazelDepGraphValue.getCanonicalRepoNameLookup().get(repositoryName) == null
- maybeGetModuleExtensionForRepo(repositoryName) == null

The conditions seem mostly fine, providing we can't use `bazel_tools` for either a module name or module extension within bzlmod. Is this true?
In case it can be used, you're back forming a cycle as soon as the new `bazel_tools` uses `builtins`. Or even worse, you don't allow this new repo to be used from anywhere (even when it should be locally available).

What I'm also concerned is, this doesn't seem to handle the same cycle when `not enableBzlMod`.

Also, because the cycle is started from @_builtins repository mapping request. I think this would be better put into `BzlloadFunction.getRepositoryMapping`, so that it covers only @_builtins -> @bazel_tools edge.

In that function it might even be clearer which cases need cycle-breaking.

(I'm not an expert in SkyFrame, I've written what I think and not everything might be completely accurate)

**Review by comius — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/devtools/build/lib/skyframe/RepositoryMappingFunction.java
Comment (comius) [reply src/main/java/com/google/devtools/build/lib/skyframe/RepositoryMappingFunction.java:139]: I'd also suggest adding a test for the minimal reproduction in the bug, having bzlmod enabled and disabled.

Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/skyframe/RepositoryMappingFunction.java:139]: Hang on, I still don't understand. the branch on line 119 should include `@bazel_tools`, as `bazel_tools` is a builtin module ("builtin" as in "shipped with Bazel", not as in `@_builtins`). So this check on line 139 should never occur.

(In other words, from what Ivo wrote, the fourth bullet point should not hold.)

Alternatively, the question is -- how come this cycle wasn't a problem before? How does injection builtins for WORKSPACE-loaded .bzl files affect the repo mapping calculation of `@bazel_tools` when Bzlmod is enabled?

File: src/main/java/com/google/devtools/build/lib/packages/BazelStarlarkEnvironment.java
Comment (benjaminp) [reply src/main/java/com/google/devtools/build/lib/packages/BazelStarlarkEnvironment.java:379]: Yes—I think what we should do is remove support for disabling builtins injections, since that can't possibly result in a working Bazel anymore.

**Review by comius — APPROVED** [review]

File: src/main/java/com/google/devtools/build/lib/skyframe/RepositoryMappingFunction.java
Comment (comius) [reply src/main/java/com/google/devtools/build/lib/skyframe/RepositoryMappingFunction.java]: This looks good to me now. @Wyverald any concerns?

**Comment by sgowroji** [comment]
Hi @comius, Since I can see that this PR has been approved, please let me know whether I should proceed with importing it. Thanks!

**Comment by comius** [comment]
> Hi @comius, Since I can see that this PR has been approved, please let me know whether I should proceed with importing it. Thanks!

Please wait for @Wyverald's approval as well. Thanks.

**Review by Wyverald — APPROVED** [review]

**Pull request closed without merging**
